### PR TITLE
feat: add prometheus metric

### PR DESCRIPTION
**Pull request by fatelei**
ref: https://github.com/cectc/dbpack/issues/25

### Ⅰ. Describe what this PR did

add prometheus metric for dbpack. opentelemetry metric currently unstable 
image

### Ⅱ. Does this pull request fix one issue?

#25 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 

no need tests

### Ⅳ. Describe how to verify it

in my computer, i verify it

### Ⅴ. Special notes for reviews

connection metric filter config
image

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -18,6 +18,10 @@ package main
 
 import (
 	"context"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"net"
+	"net/http"
 	"os"
```
Comment (dk-lockdown) [inline cmd/cmd.go:24]: pls sort the import block. [Contributing](https://github.com/CECTC/dbpack/blob/dev/CONTRIBUTING.md) there is a guide.

File: cmd/cmd.go [hunk 2]
```diff
@@ -171,6 +184,19 @@ func init() {
 	rootCommand.AddCommand(startCommand)
 }
 
+func initStatus(lis net.Listener) {
+	mux := http.NewServeMux()
```
Comment (dk-lockdown) [inline cmd/cmd.go:191]: initMetrics?

**Comment by fatelei** [comment]
I have read the CLA Document and I hereby sign the CLA

File: pkg/filter/dt/filter_mysql.go [hunk 3]
```diff
@@ -95,49 +95,62 @@ func (f *_mysqlFilter) GetName() string {
 }
 
 func (f *_mysqlFilter) PreHandle(ctx context.Context, conn proto.Connection) error {
+	var err error
 	bc := conn.(*driver.BackendConnection)
 	commandType := proto.CommandType(ctx)
 	if commandType == constant.ComStmtExecute {
 		stmt := proto.PrepareStmt(ctx)
 		if stmt == nil {
+			mysqlFilterErrorCount.WithLabelValues("unknown", "pre").Inc()
```
Comment (dk-lockdown) [inline pkg/filter/dt/filter_mysql.go]: metrics should create a new filter, `/pkg/filter/dt/filter_mysql` only process distributed transaction problem.

Comment (dk-lockdown) [reply pkg/filter/dt/filter_mysql.go]: label can be db name、table name、sql type(select、update、delete、insert)

**Comment by fatelei** [comment]
the test metric is below:

image

File: pkg/proto/interface.go [hunk 4]
```diff
@@ -72,6 +72,12 @@ type (
 		PostHandle(ctx context.Context, result Result) error
 	}
 
+	DBConnectionMetricFilter interface {
+		Filter
+		PreHandle(ctx context.Context, conn Connection) error
+		PostHandle(ctx context.Context, result Result, conn Connection) error
+	}
+
```
Comment (dk-lockdown) [inline pkg/proto/interface.go]: this interface is unnecessary! metricsFilter can be asserted as DBConnectionPreFilter and DBConnectionPostFilter

File: cmd/cmd.go [hunk 5]
```diff
@@ -18,13 +18,17 @@ package main
...
+	"net/http"
 	"os"
 	"os/signal"
 	"syscall"
 	"time"
 
 	_ "github.com/go-sql-driver/mysql"
 	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"github.com/spf13/cobra"
```
Comment (dk-lockdown) [inline cmd/cmd.go:32]: pls sort import block

File: pkg/filter/metrics/connection_metric.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+type ConnectionMetricFactory struct{}
+
+func (factory *ConnectionMetricFactory) NewFilter(config map[string]interface{}) (proto.Filter, error) {
+	connectionFilterExecDuration := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "dbpack",
+			Subsystem: "connection_filter",
+			Name:      "execute_duration",
```
Comment (dk-lockdown) [inline pkg/filter/metrics/connection_metric.go]: subsystem should be `connection`

Comment (dk-lockdown) [reply pkg/filter/metrics/connection_metric.go]: pls change name to `execute_latency `

File: pkg/filter/metrics/connection_metric.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		case *ast.DeleteStmt:
+			command = "delete"
+		case *ast.InsertStmt:
+			command = "insert"
+		case *ast.UpdateStmt:
+			command = "update"
+		case *ast.SelectStmt:
+			command = "select"
+		}
+		f.connectionFilterExecDuration.WithLabelValues(conn.DataSourceName(), command).Observe(time.Since(startAt).Seconds())
```
Comment (dk-lockdown) [inline pkg/filter/metrics/connection_metric.go]: pls add a label `CommandType`, such as ComQuery、ComPrepareExecute

**Comment by fatelei** [comment]
ConnectionMetricFilter should be add first position of filters array

**Review by dk-lockdown — APPROVED** [review]
lgtm

File: pkg/filter/metrics/connection_metric.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+const (
+	connectionMetricFilterName = "ConnectionMetricFilter"
+)
+
+func init() {
+	filter.RegistryFilterFactory(connectionMetricFilterName, &ConnectionMetricFactory{})
+}
+
+type ConnectionMetricFactory struct{}
+
```
Comment (dk-lockdown) [inline pkg/filter/metrics/connection_metric.go:41]: this factory no need to be exported.

File: doc/metric.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Metric
+
+dbpack using prometheus metric, when start dbpack, dbpack will start prom http server to let prometheus scrap the metric.
```
Comment (bohehe) [inline doc/metric.md:3]:
```suggestion
dbpack uses prometheus metric, when starting dbpack, dbpack will start prom http server to let prometheus collect the metric.
```

**Review by bohehe — APPROVED** [review]

**Review by bohehe — APPROVED** [review]
LGTM

**Comment by zhu733756** [comment]
LGTM

File: cmd/cmd.go [hunk 10]
```diff
@@ -171,6 +185,19 @@ func init() {
 	rootCommand.AddCommand(startCommand)
 }
 
+func initMetrics(lis net.Listener) {
```
Comment (zhu733756) [inline cmd/cmd.go]: shall we add a context here?

**Comment by fatelei** [comment]
file formatted, i add `go fmt` on file save

**Pull request merged**